### DISCUSSION TITLE
Quay.io authentication problem

### DISCUSSION THREAD

**Discussion by FernandesMF** [renovatebot/renovate#29682] [Request Help]
### What would you like help with?

I would like help with my configuration

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

GitLab, Renovate 37.407.1-custom

### Please tell us more about your question or problem

Hi there!

I'm having trouble making renovate authenticate to quay.io. I keep getting 401 responses from the requests to quay.io.

I have searched other similar issues and tried a few things in the discussions, but so far i couldn't manage to solve this...
Could you help me figure what i might be doing wrong?

*What i tried so far*: I made a simple repo in gitlab with only one dependency, a custom image in quay.io. I run Renovate against this repo, and the problem happens when it tries to query quay.io for newer versions of my container image.

### Logs (if relevant)

<details><summary>Logs</summary>

When Renovate tries to fetch a new version of the image, it gets a 401 error when querying quay:

```
TRACE: HTTP request: GET https://quay.io/api/v1/repository/<myusername>/<myreponame>/tag/?limit=100&page=1&onlyActiveTags=true (repository=<myusername>/<myreponame>)
DEBUG: hostRules: applying Basic authentication for quay.io (repository=<myusername>/<myreponame>)
TRACE: Applying Basic authentication (repository=<myusername>/<myreponame>)
       "url": "https://quay.io/api/v1/repository/<myusername>/<myreponame>/tag/?limit=100&page=1&onlyActiveTags=true"
TRACE: No throttle (repository=<myusername>/<myreponame>)
       "host": "quay.io"
TRACE: No concurrency limits (repository=<myusername>/<myreponame>)
       "host": "quay.io"
TRACE: got request (repository=<myusername>/<myreponame>)
       "url": "https://quay.io/api/v1/repository/<myusername>/<myreponame>/tag/?limit=100&page=1&onlyActiveTags=true",
       "options": {
         "method": "get",
         "context": {"hostType": "docker"},
         "retry": {"calculateDelay": "[function]", "limit": 2, "maxRetryAfter": 0},
         "hostType": "docker",
         "responseType": "json",
         "headers": {
           "accept": "application/json",
           "user-agent": "RenovateBot/37.407.1-custom (https://github.com/renovatebot/renovate)",
           "authorization": "***********"
         },
         "hooks": {"beforeRedirect": ["[function]"]},
         "readOnly": true,
         "timeout": 60000
       }
DEBUG: GET https://quay.io/api/v1/repository/<myusername>/<myreponame>/tag/?limit=100&page=1&onlyActiveTags=true = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=401 retryCount=0, duration=37) (repository=<myusername>/<myreponame>)
```
I tested the credentials with podman, and they worked fine:
```
> podman login quay.io --log-level=trace

Username: <myusername>+
Password: 
DEBU[0028] Looking for TLS certificates and private keys in /etc/docker/certs.d/quay.io 
DEBU[0028] GET https://quay.io/v2/                      
DEBU[0029] Ping https://quay.io/v2/ status 401          
DEBU[0029] GET https://quay.io/v2/auth?account=<myusername>%2B&service=quay.io 
DEBU[0029] Increasing token expiration to: 60 seconds   
DEBU[0029] GET https://quay.io/v2/                      
DEBU[0029] Stored credentials for quay.io in credential helper containers-auth.json 
Login Succeeded!
```
I noticed that podman is querying the v2 api version:
```
> podman pull quay.io/<myusername>/<myreponame>:v1.0 --log-level=trace

DEBU[0000] Looking for TLS certificates and private keys in /etc/docker/certs.d/quay.io 
DEBU[0000] GET https://quay.io/v2/                      
DEBU[0000] Ping https://quay.io/v2/ status 401          
DEBU[0000] GET https://quay.io/v2/auth?account=<myusername>%2B&scope=repository%3A<myusername>%2F<myreponame>%3Apull&service=quay.io 
DEBU[0001] Increasing token expiration to: 60 seconds   
DEBU[0001] GET https://quay.io/v2/<myusername>/<myreponame>/manifests/v1.0
```
while Renovate is using the v1 api.

Please let me know if there is any further information that could be helpful. Thanks!

</details>

**Comment by rarkins** [comment]
I tested with a public quay.io image. It was successful and resulted in these API calls:

```
         "https://quay.io/api/v1/repository/openshift-release-dev/ocp-release/tag/": {
           "GET": {"200": 20}
         },
         "https://quay.io/v2/": {"GET": {"401": 1}},
         "https://quay.io/v2/auth": {"GET": {"200": 1}},
         "https://quay.io/v2/openshift-release-dev/ocp-release/blobs/sha256:4feaa1f8bea64966cdbbf25779bd577383aa6e9df7ad9806851a07a4a039ca7e": {
           "GET": {"200": 1}
         },
         "https://quay.io/v2/openshift-release-dev/ocp-release/manifests/4.17.0-ec.0-aarch64": {
           "GET": {"200": 1}
         }
```

When I removed the conditional quay.io logic we have to force v1, the PR created in the first run is autoclosed, even though the api requests seem to have succeeded:

```         "https://quay.io/v2/": {"GET": {"401": 1}},
         "https://quay.io/v2/auth": {"GET": {"200": 1}},
         "https://quay.io/v2/openshift-release-dev/ocp-release/blobs/sha256:c1fa36ed5a4c243e5867dd411159d54b3015623af9de5295a032bc3c8c6309c8": {
           "GET": {"200": 1}
         },
         "https://quay.io/v2/openshift-release-dev/ocp-release/manifests/4.9.10-ppc64le": {
           "GET": {"200": 1}
         },
         "https://quay.io/v2/openshift-release-dev/ocp-release/tags/list": {
           "GET": {"200": 20}
         }
```

Looking at the results for https://quay.io/v2/openshift-release-dev/ocp-release/tags/list?n=10000 it seems that it returns 50 tags at a time, from oldest to newest.	

When I modify the maximum API page limit from 20 to 2000 and retry, Renovate needs to fetch 182 pages to get the full tags list:

```
         "https://quay.io/v2/openshift-release-dev/ocp-release/tags/list": {
           "GET": {"200": 182}
         }
```

From this, I'm concluding that quay's problem with pagination (limiting to only 50 tags per page, returning oldest first) is why we use the v1 API for tag listing instead of v2.

So @FernandesMF can you find a way to curl `https://quay.io/api/v1/repository/<myusername>/<myreponame>/tag/?limit=100&page=1&onlyActiveTags=true` successfully?

**Reply by viceice** [reply]
@rarkins yes, that's why we use the quay rest API instead of the docker API

**Reply by viceice** [reply]
https://github.com/renovatebot/renovate/issues/9604 is maybe also relevant

**Reply by FernandesMF** [reply]
Thanks for your reply!

I tried to curl the address you suggested, @rarkins, but I still got a 401:
```
> curl -u <user>:<pw> https://quay.io/api/v1/repository/<myusername>/<myreponame>/tag/\?limit\=100\&page\=1\&onlyActiveTags\=true
{"detail": "Requires authentication", "error_message": "Requires authentication", "error_type": "invalid_token", "title": "invalid_token", "type": "https://quay.io/api/v1/error/invalid_token", "status": 401}
```
Do you have any hint of what I could try differently?

**Reply by viceice** [reply]
you first need to get the url for getting an auth token, then you need to send your basic auth creds to that url to get the token. after all you can use the token as bearer auth to do authenticated requests 

https://distribution.github.io/distribution/spec/auth/token/

**Reply by FernandesMF** [reply]
Hi! Sorry for the long time to reply, these last days were a bit hectic.

I assembled the following script to try the authentication loop:
```
set -euo pipefail
 
USER=<username>
PW=<password>
REPO=<repository>
 
TOKEN=$(curl -sLkv -u $USER:$PW "https://quay.io/v2/auth?service=quay.io&scope=repository:$REPO:pull" | jq '.token' -r)

curl -skL -H "Authorization: Bearer $TOKEN" https://quay.io/api/v1/repository/$REPO/tag/\?limit\=100\&page\=1\&onlyActiveTags\=true
```
(I researched a few other discussions here in github, and I saw this url 'https://quay.io/v2/auth?service=quay.io&scope=repository:$REPO:pull', which I'm assuming is the one to used for token requests)

The token request seems to return a valid token. However, the last curl returns the following error:
```
{"error": "Issuer quay not configured"}
```
which I'm still not understanding completely 😅 

Perhaps the token url is not quite right?

Thanks again for the help!

Edit: I also tried using the token url `https://quay.io/v2/auth?account=$USER&scope=repository:$REPO%3Apull&service=quay.io`, which is the one I got from the podman trace duirng the pull, and the issue persists.

**Reply by FernandesMF** [reply]
Hi @viceice, @rarkins!

Is there anything you would like me to try (a command, a task, fetch some logs/info, etc) to help figure this issue out? Or perhaps should I file another ticket with a different iissue type?

Please let me know if you are awaiting something (result, test, info, etc) from me. Thanks!

**Reply by rarkins** [reply]
We need a sequence of `curl` commands which can be used to fetch the tags. Without that, there's nothing to do. Please don't keep @'ing us

**Comment by viceice** [comment]
please try this 

https://github.com/renovatebot/renovate/discussions/10188#discussioncomment-792567

**Reply by FernandesMF** [reply]
Thanks for the reply, @viceice!

I gave a try using the config options in the link you sent, and I still got a 401.

Here is my config:
```
"hostRules": [
  {
     "matchHost": "docker.io",
     "username": "<dockerhub_username>",
     "password": "***********"
  },
  {
     "hostType": "docker",
     "matchHost": "https://quay.io/api/v1/",
     "token": "***********"
  }
]
```

**Reply by FernandesMF** [reply]
Hi @viceice!

Is there anything you would like me to try (a command, a task, fetch some logs/info, etc) to help figure this issue out? Or perhaps should I file another ticket with a different iissue type?

Please let me know if you are awaiting something (result, test, info, etc) from me. Thanks!
